### PR TITLE
Rewrite library view

### PR DESCRIPTION
**Pull request by aweebs**
Generally works the same as the previous page, with the bonus of having a watching on the viewId parameter for when libraries change.

### CODE REVIEW

File: frontend/src/components/Buttons/FilterButton.vue
Comment (ferferga) [inline frontend/src/components/Buttons/FilterButton.vue:115]: This check is repeated. Can't we move it to a computed property?

File: frontend/src/components/Buttons/FilterButton.vue
Comment (ferferga) [inline frontend/src/components/Buttons/FilterButton.vue:266]: Fix the ``refesh`` typo above this please (GH doesn't let me select that)

File: frontend/src/components/Buttons/FilterButton.vue
Comment (ferferga) [inline frontend/src/components/Buttons/FilterButton.vue:310]: The logic behind the library/filter/sort buttons have always seemed so tangled to me that I never tried to understood it properly (not your fault ofc, in fact you're brave by trying to fix past messes!)
However, better be safe than worry: are you sure we don't need to call ``refreshItems`` at mount? It's truly only needed when filters are applied?

In case you discover we need it at mount, make this watcher immediate please.

File: frontend/src/components/Buttons/SortButton.vue
Comment (ferferga) [inline frontend/src/components/Buttons/SortButton.vue:24]: Maybe this can be computed too?

File: frontend/src/components/Buttons/TypeButton.vue
Comment (ferferga) [inline frontend/src/components/Buttons/TypeButton.vue:5]: Why? You're cluttering the setup with this change imo. At the past, @ThibaultNocchi and me discussed about the different methods of importing icons and we agreed that having icons both defined as prop or as components is fine, since the templates can get very cluttered with how Vuetify uses icon as slots. However, in this case, we take even more space (when icons use around 1:1 of the space)

File: frontend/src/components/Buttons/TypeButton.vue
Comment (ferferga) [inline frontend/src/components/Buttons/TypeButton.vue:45]: Thank you for replacing the remnants of not having ``BaseItemKind`` properly typed!

File: frontend/src/utils/__tests__/items.spec.ts
Comment (ferferga) [inline frontend/src/utils/__tests__/items.spec.ts:49]: I don't care about getting this change in or out, but I was actually thinking on removing all the test files before merging and do proper testing from the start using Vitest or Cypress (which was the next major thing I wanted to tackle after Vue 3).

If you are going to be around and want to help in the testing suite, I'll leave this up to you (maybe you prefer to incrementally refactor, instead of a complete redo?) :).

File: frontend/src/pages/library/_viewId.vue
Comment (ferferga) [inline frontend/src/pages/library/_viewId.vue:37]: Same as above, maybe we can have this computed?

File: frontend/src/pages/library/_viewId.vue
Comment (ferferga) [inline frontend/src/pages/library/_viewId.vue:80]: Please, let's try to keep all the concerns together. The order is not that relevant to me, but ideally all definitions (refs or computed), functions and then watchers at the bottom.

Why? Besides looking more clear and easier to find them if we stick to some convention at this, we avoid ``ReferenceError`` when an immediate watcher could be fired before a variable is assigned.

Also, maybe this can be computed too?

File: frontend/src/pages/library/_viewId.vue
Comment (ferferga) [inline frontend/src/pages/library/_viewId.vue:85]: Make this an immediate watcher

File: frontend/src/pages/library/_viewId.vue
Comment (ferferga) [inline frontend/src/pages/library/_viewId.vue:186]: Said the ``ReferenceError`` thing above and then I discover this function is defined after the watcher :)
See what I mean?

**Review by ferferga — CHANGES_REQUESTED** [review]

File: frontend/src/components/Buttons/TypeButton.vue
Comment (aweebs) [reply frontend/src/components/Buttons/TypeButton.vue:5]: I like the idea of the code part of the template explicitly declaring all the requirements that the template will need, rather than have "global" variables in the template. But if there's been a decision to not do this then I will revert it.

File: frontend/src/utils/__tests__/items.spec.ts
Comment (aweebs) [reply frontend/src/utils/__tests__/items.spec.ts:49]: I'd prefer to keep it in, I'm a fan of incremental changes in the right direction. I think having these old test cases around will help when the test suite is upgraded. I hope I'll still be around to help with that!

File: frontend/src/pages/library/_viewId.vue
Comment (aweebs) [reply frontend/src/pages/library/_viewId.vue:80]: I'll re-organize like that. I was organizing these by the parts that they impacted.

Not sure this can be computed though since there's an assignment as part of it which means it's a side effect and should be a watch.

File: frontend/src/components/Buttons/TypeButton.vue
Comment (ferferga) [reply frontend/src/components/Buttons/TypeButton.vue:5]: @aweebs I thought this was the reasoning behind, so somehow we agreed beforehand :).

And yes, I agree some things like this are confusing in Vue (like the ref unwrapping). No strong feelings about it, just wanted to be sure if it was conscious and not because you thought it wasn't working, for instance.

File: frontend/src/utils/__tests__/items.spec.ts
Comment (ferferga) [reply frontend/src/utils/__tests__/items.spec.ts:49]: @aweebs Okay. Be aware anyway that removing them will still make them available in history, so we're not losing them really. It's just a matter of deciding between how easy it will be to review new files vs or incrementally modify them (because if the modifications are going to be huge, maybe it's not worth it).

I looked at vitest API and it looks really similar to jest, so I'm not sure what to think (this is why I'm asking :P)

File: frontend/src/pages/library/_viewId.vue
Comment (ferferga) [reply frontend/src/pages/library/_viewId.vue:80]: @aweebs My point is to have a computed ``libName`` for the if checking and then assign it like ``route.meta.title = libName.value``.

Though now that I'm thinking about this... I'm not sure if maybe we don't need the watcher or the computed property at all since lib should already be reactive? Can you check it please?

Comment (aweebs) [reply frontend/src/pages/library/_viewId.vue:80]: Using just an assignment without a computed value doesn't work. Neither does assigned it to a computed value like so:
```
const title = computed(() => {
  if (
    library.value &&
    library.value.Name &&
    library.value.Type &&
    validLibraryTypes.includes(library.value.Type)
  ) {
    return library.value.Name;
  }
});

route.meta.title = title.value;
```

I think this will need to stay as a watcher

Comment (aweebs) [reply frontend/src/pages/library/_viewId.vue:80]: I'm going to simplify this too so that it always just assigned the Name, since only libraries should be loaded here I don't think there's any point in checking if it's a valid library type.

**Comment by aweebs** [comment]
I think the feedback has largely been addressed, the last thing that's needed before merge is to debounce the item search and also check the filter type button since I think it's triggering another search once the filters load.

File: frontend/src/pages/library/_viewId.vue
Comment (ferferga) [reply frontend/src/pages/library/_viewId.vue]: @aweebs Fine then.

The check could be moved to the middleware level instead, it doesn't hurt really. Doing that in this PR is up to you though.

**Comment by aweebs** [comment]
> @aweebs This is good to go as soon as the unused debounce import is removed and the comment is fixed imo. The refresh is probably triggered by the type button watcher that I mentioned above.

@ferferga I've removed the unused import for now and update the multiline comment to docblock which I think was the other blocker. We can add debouncing at another time.

**Review by ferferga — APPROVED** [review]

**Pull request merged**
